A type t, equal, show and pp

Leonidas-from-XIV opened the pull request:
Closes #71.

I think most of it is pretty obvious except for how equality on `Assoc` would be handled. Since keys can repeat in that association list, I implemented it in a way where same keys must have the same value (thus a custom `mem`-like function). It is not optimal, since that comparison is `O(n^2)`, but I decided to submit it anyway to have a discussion on whether these semantics are what we want to have to begin with. There are some obvious optimizations but let's take this as starting point.

NathanReb commented:
Thanks for that!

I restarted the failed job and I'll have a look ASAP.

Leonidas-from-XIV commented:
Looks like it worked. I haven't included `compare` because I don't have too much use for it personally, but if there is interest I can also add that, by the magic of "stealing whatever ppx_deriving generates".

Leonidas-from-XIV commented:
While working on another PR I realized there is `sort` which might actually be the better solution to determining whether two `` `Assocs`` are equivalent, especially since the `sort` logic exists and due to its existence is the "canonical" way of ordering keys in an `` `Assoc``. Let me know what you think.

NathanReb commented on lib/type.ml:
```diff
@@ -52,6 +52,12 @@ All possible cases defined in Yojson:
 *)
+
+type json = t
+(**
```
We should deprecate the alias so we can get rid of it when we bump the major version.

Leonidas-from-XIV commented on lib/monomorphic.mli:
```diff
@@ -0,0 +1,8 @@
+  (** Convert value to string, useful for debugging *)
+
+val equal : t -> t -> bool
+  (** [equal a b] is the monomorphic equality *)
```
I added some description on the behaviour. Can certainly be improved but at least now it should be somewhat clearer.

Leonidas-from-XIV commented:
I will add unit tests to it (and actually it is way nicer now, since I have an equality and a pretty printer that I can use in the tests), but since that takes a bit of time I submitted the other changes that you requested since they were quick to do.

Leonidas-from-XIV commented:
After adding the deprecation annotation I had to change all occurrences of `json` to `t` to not trigger deprecation warnings ourselves.

Leonidas-from-XIV commented on lib/monomorphic.ml:
```diff
@@ -0,0 +1,161 @@
+    | `Stringlit a, `Stringlit b -> a = b
+#endif
+    | `Assoc xs, `Assoc ys ->
+      let xs, ys = match sort (`Assoc xs), sort (`Assoc ys) with
```
That's a good point. The issue I see here is that just sorting that way will cause problems with duplicate keys, since `{"a": 42, "a": 23}` might or might not be equal depending on whether it will reorder the duplicate pair in the alist. And sorting "properly" requires to sort by key-first value-second, but for on the value we'd need `compare`.

Maybe I should just bite the bullet and implement compare and then sort in a sensible way, what do you think?

NathanReb replied:
Although Yojson do accept such values, given what the [JSON RFC](https://tools.ietf.org/html/rfc8259#section-4) specifies I think we shouldn't bother with key duplicates and go with a sort on keys solely, cf. my suggestion on the doc comment below.

Because `compare` will rely on a sort itself anyway I don't think implementing it will really help figure this out. We will just migrate this issue from being `equal`s problem to being `compare`'s one. It also feels weird to have `equal` rely on `compare` in any other way than just using it straight and comparing the result to `0`.

I'd say let's not support duplicate in `equal` for now

Leonidas-from-XIV replied:
Okay, I implemented it with `stable_sort` which I think is a decent compromise between practicality, performance and correctness.

Leonidas-from-XIV commented on lib/monomorphic.ml:
```diff
@@ -0,0 +1,161 @@
+        | `Assoc xs, `Assoc ys -> xs, ys
+        | _ -> raise (Failure "Sorting Assoc returned non-Assoc")
+      in
+      (match List.fold_left2 (fun acc (key, value) (key', value') ->
```
It's a good point, fixed.

NathanReb replied:
Sorry I missed it but this also applies to the `List` case, could you fix that one as well? Thanks!

Leonidas-from-XIV commented on lib/monomorphic.mli:
```diff
@@ -0,0 +1,12 @@
+  (** [equal a b] is the monomorphic equality.
+      Determines whether two JSON values are considered equal. In the case of
+      JSON objects, the order of the keys does not matter. In case of duplicate
+      keys, these have to exist in both objects to be considered equal.
```
Yes, that's exactly the point I'm discussing.

Correctness in this context is an odd thing, because whether duplicate keys are valid or not does not seem to be clear. Some implementations pick the key parsed last, so in such case `{"a": 42, "a": 23}` would not be equal to `{"a": 23, "a": 42}` in any case because the parsed representation doesn't even include the duplicate. Yojson does allow that and also writes them out in the same order.

The more I think about it the more it seems to me like we should consider them equal if they are duplicated only if they are in the same order, which I guess is exactly what we get with `stable_sort`.

Leonidas-from-XIV commented:
I added some unit tests, though I didn't bother to test the non-standard extensions to JSON. It uses alcotest which I added as a test-only dependency.

I think the CI scripts will run the tests automatically, though we might possibly disable them on all but one build because running the tests takes a while because it rebuilds it with test dependencies.
